services/web/server/src/simcore_service_webserver/projects/_db_utils.py

```diff
@@ -370,3 +377,98 @@ async def _get_project(
             project["tags"] = tags
 
         return project
+
+
+def update_workbench(old_project: ProjectDict, new_project: ProjectDict) -> ProjectDict:
+    """any non set entry in the new workbench is taken from the old one if available
+
+    Raises:
+        ProjectInvalidUsageError: it is not allowed to add/remove nodes
+
+    Returns:
+        udpated project
+    """
+
+    old_workbench: dict[NodeIDStr, Any] = old_project["workbench"]
+    updated_project = deepcopy(new_project)
+    new_workbench: dict[NodeIDStr, Any] = updated_project["workbench"]
+
+    if old_workbench.keys() != new_workbench.keys():
+        # it is forbidden to add/remove nodes here
+        raise ProjectInvalidUsageError
+    for node_key, node in new_workbench.items():
+        old_node = old_workbench.get(node_key)
+        if not old_node:
+            continue
+        for prop in old_node:
+            # check if the key is missing in the new node
+            if prop not in node:
+                # use the old value
+                node[prop] = old_node[prop]
+    return updated_project
+
+
+def patch_workbench(
+    project: ProjectDict,
+    *,
+    new_partial_workbench_data: dict[NodeIDStr, Any],
+    allow_workbench_changes: bool,
+) -> tuple[ProjectDict, dict[NodeIDStr, Any]]:
+    """patch the project workbench with the values in new_partial_workbench_data
+
+    - Example: to add a node: ```{new_node_id: {"key": node_key, "version": node_version, "label": node_label, ...}}```
+    - Example: to modify a node ```{new_node_id: {"outputs": {"output_1": 2}}}```
+    - Example: to remove a node ```{node_id: None}```
+
+    Arguments:
+        project -- _description_
+        new_partial_workbench_data -- _description_
+        allow_workbench_changes -- _description_
+
+    Raises:
+        ProjectInvalidUsageError: if allow_workbench_changes is False and user tries to add/remove nodes
+        NodeNotFoundError:
+
+    Returns:
+        patched project and changed entries
+    """
+
+    changed_entries = {}
+    for (
+        node_key,
+        new_node_data,
+    ) in new_partial_workbench_data.items():
+        current_node_data: dict[str, Any] | None = project.get("workbench", {}).get(
+            node_key
+        )
+
+        if current_node_data is None:
+            if not allow_workbench_changes:
+                raise ProjectInvalidUsageError
+            # if it's a new node, let's check that it validates
+            try:
+                Node.parse_obj(new_node_data)
+                project["workbench"][node_key] = new_node_data
+                changed_entries.update({node_key: new_node_data})
+            except ValidationError as err:
+                raise NodeNotFoundError(project["uuid"], node_key) from err
```

Comment:
I think not found is misleading here. I rather add a new error to indicate that the node data update is invalid

Reply:
I agree but it did not change with my PR. as discussed this remains as it is for now.